pkg/util/log/log.go

```diff
@@ -38,24 +41,64 @@ func InitLogger(cfg *server.Config, reg prometheus.Registerer) {
 type prometheusLogger struct {
 	logger      log.Logger
 	logMessages *prometheus.CounterVec
+	logFlushes  prometheus.Histogram
+
+	useBufferedLogger bool
+	useSyncLogger     bool
 }
 
 // newPrometheusLogger creates a new instance of PrometheusLogger which exposes
 // Prometheus counters for various log levels.
-func newPrometheusLogger(l logging.Level, format logging.Format, reg prometheus.Registerer) log.Logger {
-	logger := log.NewLogfmtLogger(log.NewSyncWriter(os.Stderr))
+func newPrometheusLogger(l logging.Level, format logging.Format, reg prometheus.Registerer, buffered bool, sync bool) log.Logger {
+
+	// buffered logger settings
+	var (
+		logEntries    uint32 = 256                    // buffer up to 256 log lines in memory before flushing to a write(2) syscall
+		logBufferSize uint32 = 10e6                   // 10MB
```

Comment:
what happens if 256 entries goes over 10MB? Will the buffer also flush on size?

Reply:
The preallocated buffer is just to prevent slice resizing, and because it's reused it prevents unnecessary GC.
If 10MB is exceeded then the buffer's underlying slice will grow, so nothing will really happen - it won't flush on size.